Middleware: Add Password Authentication with JWT and Bcrypt

TheCharlatan opened the pull request:
RPC calls to the middleware need to be authenticated after this pull request is merged. To authenticate, a client first calls the UserAuthenticate method. On first setup of the base, the username and password are set to 'admin' and 'ICanHasPasword?'. The passed in password is then compared with a bcrypt hash of the password in redis. Redis stores a JSON 'auth' structure for this that can be accessed with the 'auth' key. The structure maps between a username and its password and role.

Once the user's password is successfully compared he is issued a JWT (JSON Web Token). He then needs to include this token in every rpc call to be successfull. The rpc methods return an error otherwise.

To change the password, the UserChangePassword method first checks the length of the new password. Then it fetches the 'auth' structure from redis, checks if the username exists and the provided current password is correct. Once done the new password is hashed and encoded with bcrypt and finally saved to redis.

A lot of unit tests had to be adapted for this change. Most importantly the rpc tests now only test against a mocked version of the middleware, which was generated with mockery. 

I removed some units tests from the rpcserver test that I thought rather test the middleware's behaviour, than the rpcserver's.

![AuthenticationDiagram](https://user-images.githubusercontent.com/8421793/66650911-06e98d00-ec32-11e9-8d45-8b9844f17939.png)

0xB10C commented on middleware/src/rpcmessages/errorcodes.go:
```diff
@@ -136,9 +138,31 @@ const (
 
-	// ErrorDummyPasswordTooShort is thrown if the provided password is too short.
-	ErrorDummyPasswordTooShort ErrorCode = "DUMMY_CHANGEPASSWORD_TOO_SHORT"
+	// ErrorAuthenticationPasswordIncorrect is thrown if the dummy authentication is not successful.
```
nit: remove `dummy` from comment

0xB10C commented on middleware/src/rpcmessages/errorcodes.go:
```diff
@@ -136,9 +138,31 @@ const (
+	ErrorAuthenticationFailed ErrorCode = "AUTHENTICATION_FAILED"
+
+	// ErrorAuthentcationUsernameNotFound is thrown if the given username does not exist.
+	ErrorAuthenticationUsernameNotFound ErrorCode = "AUTHENTICATION_USERNAME_NOEXIST"
```
Should we really differentiate between a failed authentication and an authentication where the username does not exist? It helps the user when he e.g. mistypes the username, but also enables an attacker to brute force usernames. IIRC it's common practice (unless you are Facebook or Google) to not reveal when a username doesn't exist.

TheCharlatan replied:
Currently we only have a single "admin" user, so I'll leave this distinction for now to make it easier for us. Once we support multiple users and the adding of additional users, we can work over it again. I'll add a TODO to indicate this.

0xB10C commented on middleware/src/authentication/authentication.go:
```diff
@@ -0,0 +1,95 @@
+	// generate random string with 32 bytes of entropy and use it as the signing key
+	JwtKey, err := jwtAuth.generateRandomString(32)
+	if err != nil {
+		log.Println("could not get enough entropy to generate key")
```
>```go 
>// GenerateRandomBytes returns securely generated random bytes. It will return an error
>// if the system's secure random number generator fails to function correctly, in which
>// case the caller should not continue.
>```

Did I miss something or do you just continue if there is not enough entropy?

TheCharlatan replied:
If you check the entire NewMiddleware method, I am not returning from the setup if any error occurs. This is obviously bad. I think we should panic if these setup functions fail, what do you think? Alternatively we could also return the NewMiddleware method with an error and then exit the main process gracefully. I think in this place I prefer a panic though.

TheCharlatan replied:
I discussed with @Stadicus offline that I'll just log the error and exit with a fatal log message in the main function if there is a redis problem at the middleware startup. 
In the case of missing entropy or another problem with the jwt start I'll let the NewMiddleware function panic, since this should never happen on a device like ours.

x1ddos replied:
Please don't panic. Panic is for a programmer's (human) error. Lack of entropy is not a human error.

x1ddos replied:
Best is to return the error and let the caller handle it.

TheCharlatan replied:
Thanks Alex, dropped the panic and returned the error instead.

0xB10C commented on middleware/src/authentication/authentication.go:
```diff
@@ -0,0 +1,95 @@
+	return jwtAuth
+}
+
+// GenerateRandomBytes returns securely generated random bytes. It will return an error
```
minor nit: `GenerateRandomBytes` -> `generateRandomBytes`

0xB10C commented on middleware/src/authentication/authentication.go:
```diff
@@ -0,0 +1,95 @@
+func (jwtAuth *JwtAuth) GenerateToken(username string) (string, error) {
+	jwtStaticKey := []byte(jwtAuth.jwtKey)
+
+	// TODO: Implement proper handling of the expiration time. This is basically just a dummy value 24 hours in to the future.
```
What are your thoughts here regarding:
- When should this be done?
- What would a good expiry time?

TheCharlatan replied:
Pulling @Stadicus and @KasparEtter into this discussion.
NIST recommends an absolute maximum of 30 days for the lifetime of an authentication token. However mobile apps for example never require their users to re-authenticate manually. To enact this they use so called 'refresh tokens' which have a far longer expiry time to get a new authentication token once the authentication token has expired. This refresh token is used once and then invalidated. I don't like this though, in my eyes it just adds a layer of complexity and obscurity. It is possible to do correctly through oauth, but then requires a 3rd party to function properly. It's also a bad idea within the current architecture, since the token is no longer valid if the middleware has to restart (it generates a new jwt key on every restart). 
Since jwt makes it hard to log out users from a session, it seems to be preferable in my eyes to have a fairly low expiry time and then ask the user to re-authenticate with his password once the jwt token expires. I think a day is reasonable in this case, since we have additional security guarantees through the usage of the noise pairing. What are your opinions on this matter?

TheCharlatan replied:
After some offline discussion I'll opt to go with how it is done currently. A user will be forced to re-authenticate after 24h, notifications will still be streamed users even after their token is expired. Eventually there could be some rpc calls that are not authenticated that provide basic functionality (like an electrs proxy)

0xB10C replied:
`TODO` can be removed?

TheCharlatan replied:
Done.

0xB10C commented on middleware/src/middleware.go:
```diff
@@ -264,7 +295,10 @@ func (middleware *Middleware) BackupHSMSecret() rpcmessages.ErrorResponse {
 		}
 	}
-
+	err = middleware.redisClient.SetString(redis.BaseSetupDone, "1")
```
Is `BackupHSMSecret()` the right place for this? I don't think this is even called in the current setup Wizard. Should probably have something explicit that is called by the App once the setup can be finalized.

TheCharlatan replied:
Ah, this is a bad mistake. I need something that detects when the initial setup is done in the frontend. Should I move it to BackupSysconfig instead? @Stadicus do you have an idea what could act as a trigger to mark the base as initialized?

0xB10C replied:
We've discussed something similar this morning with @Stadicus @Tomasvrba. We don't want to restart bitcoind multiple times when applying the changes made in the setup wizard, because it takes time. One idea would be to have a RPC that finalizes the setup process, sets the redis key(s) and restarts bitcoind once.

TheCharlatan replied:
As per offline discussion I'm moving it to BackupSysconfig.

0xB10C commented on middleware/src/middleware.go:
```diff
@@ -322,59 +356,213 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
 	return rpcmessages.ErrorResponse{Success: true}
 }
 
+func (middleware *Middleware) checkMiddlewareSetup() error {
```
could be move to `util.go`

0xB10C commented on middleware/src/middleware.go:
```diff
@@ -322,59 +356,213 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+					ErrorResponse: &rpcmessages.ErrorResponse{
+						Success: false,
+						Message: "authentication unsuccessful",
+						Code:    rpcmessages.ErrorDummyAuthenticationNotSuccessful,
```
Could change from `ErrorDummyAuthenticationNotSuccessful` -> `ErrorAuthenticationNotSuccessful`

TheCharlatan replied:
Changed to ErrorInitialAuthenticationNotSuccessful

0xB10C commented on middleware/src/middleware.go:
```diff
@@ -322,59 +356,213 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+			ErrorResponse: &rpcmessages.ErrorResponse{
+				Success: false,
+				Message: "authentication failed, redis error",
+				Code:    rpcmessages.ErrorAuthenticationFailed,
```
fyi: There is a ErrorCode for a Redis Error called `ErrorRedisError`. But no strong preference on what to use here.

TheCharlatan replied:
I'd prefer to use the authenitcation error here, probably more expresive for the app in this case.

0xB10C commented on middleware/src/middleware.go:
```diff
@@ -322,59 +356,213 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+	authStructureString, err := middleware.redisClient.GetString(redis.MiddlewareAuth)
+	log.Println(authStructureString)
+	if err != nil {
+		log.Println("lel")
```
old debug logs?

0xB10C commented on middleware/src/middleware.go:
```diff
@@ -322,59 +356,213 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
-// FIXME: This is a dummy implementation of the UserChangePassword RPC call
-// This dummy method approves all passwords which are longer or equal to 8 chars
+// The function first validates the current password with redis, then replaces it with the new password.
+// Passwords need to be longer than 8 chars.
```
```suggestion
// Passwords need to be longer than or equal to 8 chars.
```

According to implementation.

0xB10C commented on middleware/src/rpcmessages/errorcodes.go:
```diff
@@ -136,9 +138,31 @@ const (
 	// ErrorDummyAuthenticationNotSuccessful is thrown if the dummy authentication is not successful.
 	ErrorDummyAuthenticationNotSuccessful ErrorCode = "DUMMY_AUTHENTICATION_NOT_SUCCESSFUL"
```
Originally dummy was meant in the way: "This is a placeholder once we have JWT Auth." Is it still needed? Could be it the case that you are using it for the "ICanHasPassword?" setup Auth?

TheCharlatan replied:
Yes, I removed the dummy prefix and renamed it to 'initial'.

0xB10C commented on middleware/src/rpcserver/mocks/Middleware.go:
```diff
@@ -0,0 +1,347 @@
+// Code generated by mockery v1.0.0. DO NOT EDIT.
```
Like that. How do we assure that the mocks are up-to-date?

TheCharlatan replied:
You need to re-run mockery. The test compilation will fail if they do not implement the full interface.

TheCharlatan replied:
Run `mockery -name Middleware` in the rpcserver directory. There are go:generate instructions for this in the rpcserver_test file. I also added them to the rpcserver file. In case your ide/editor is not configured to rerun `go generate` before every compilation run, you can execute the mockery command manually.

0xB10C replied:
Shouldn't that be included in the Makefile? Not good to be dependent on  IDE/editor config when (newly starting) developing the Middleware.

TheCharlatan replied:
Yeah, the more tooling we can hide in the Makefile, the better imo. I added it as a separate target.

0xB10C commented on middleware/src/rpcserver/rpcserver_test.go:
```diff
@@ -100,7 +131,7 @@ func TestRPCServer(t *testing.T) {
 	// The RPCs must get an argument passed.
 	// We pass a boolean to RPCs that don't need an argument.
-	dummyArg := true
+	dummyArg := rpcmessages.AuthGenericRequest{}
```
Consider renaming and changing the above comment?

TheCharlatan commented:
Force-pushed with the following changes:
- Addressed review comments (except the ones where we need further clarification)
- Renamed dummyAdminPassword to initialAdminPassword, renamed the dummy string in other places accordingly

TheCharlatan commented:
Addressed the remaining points as discussed and force pushed the changes.

0xB10C commented on middleware/src/middleware.go:
```diff
@@ -261,6 +303,11 @@ func (middleware *Middleware) BackupSysconfig() (response rpcmessages.ErrorRespo
+	err = middleware.redisClient.SetString(redis.BaseSetupDone, "1")
+	if err != nil {
+		log.Println("redis failed to set the base setup done flag. This is not a critical error, but the setup will be repeated again")
+	}
```
This should be moved eventually. I opend an issue for it https://github.com/shiftdevices/bitbox-base-internal/issues/312

Tomasvrba commented on middleware/src/middleware.go:
```diff
@@ -335,56 +381,211 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+	}
+
+	if !middleware.isMiddlewarePasswordSet {
+		if args.Username == "admin" && args.Password == initialAdminPassword {
```
What is the password is not set (`middleware.isMiddlewarePasswordSet == true`) but the wrong credentials are provided `(args.Username == "admin" && args.Password == initialAdminPassword) == false`?

I don't see an `authentication unsuccessful` error in the top if block to handle this case?

TheCharlatan replied:
Actually that entire if block can go, the rest of the code should handle it correctly. I added it to make it clear that the initial authentication is a special case, but I don't see much point in it any more. Once this is in, I'll do some more research to handle the initial case properly. I removed it now.

Tomasvrba commented on middleware/cmd/middleware/main.go:
```diff
@@ -53,7 +53,10 @@ func main() {
-	middleware := middleware.NewMiddleware(argumentMap, false)
+	middleware, err := middleware.NewMiddleware(argumentMap, false)
+	if err != nil {
+		log.Fatal("error starting the middleware, is redis connected? \nIf you are running the middleware outside of the base consider setting the redis mock flag to true.")
```
Then let's also add the flag to the top of the file to be able to start the middleware with mocked redis?

TheCharlatan replied:
Mind if I do this in a separate pr, since it is not directly related?

Tomasvrba replied:
Sure, no problem. I can just change it manually when running locally for now

Tomasvrba commented on middleware/src/middleware.go:
```diff
@@ -335,56 +381,211 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+			Message: "password change unsuccessful",
+			Code:    rpcmessages.ErrorPasswordChangeFailed,
+		}
+	}
```
You use pretty much the same block of code with `authStructureString, err := middleware.redisClient.GetString(redis.MiddlewareAuth)` and `err = json.Unmarshal([]byte(authStructureString), &usersMap)` twice. Once in `UserAuthenticate` and again here in `UserChangePassword`. Even the error messages have slightly different wording. e.g. `error getting the auth structure string from the redis client` vs. `error when getting the authentication structure from redis`

Maybe refactor this into a util function, something like `getAuthStructure` or `handleAuthStructure` that handles both the getting and the unmarshal?

TheCharlatan replied:
Great, I moved it to a separate util function.

Tomasvrba commented on middleware/src/middleware.go:
```diff
@@ -335,56 +381,211 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+		}
+	}
+
+	passwordFromStorage := usersMap[args.Username]
```
nit
This is technically not just the password from storage since you'll also get the role with `usersMap[args.Username]`
maybe do `passwordFromStorage := usersMap[args.Username].BCryptedPassword` and then you actually just have the `passwordFromStorage`?

TheCharlatan replied:
Good catch, it's actually named correctly in the UserAuthenticate function.

Tomasvrba commented on middleware/src/middleware.go:
```diff
@@ -335,56 +381,211 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
+
+	userAuthSecrets := usersMap[args.Username]
+	userAuthSecrets.BCryptedPassword = string(bcryptedPassword)
+	userAuthSecrets.Role = "admin"
```
Is the role secret? :D

TheCharlatan replied:
Depends on the ingredients, I like cinnamon in my role.

0xB10C commented on middleware/src/middleware_test.go:
```diff
@@ -236,106 +237,103 @@ func TestSetRootPassword(t *testing.T) {
-	response := testMiddleware.UserChangePassword(rpcmessages.UserChangePasswordArgs{Username: "admin", NewPassword: "abc123def"})
+	// change admin password to "abc123def", which sets isMiddlewarePasswordSet = true
+	response := testMiddleware.UserChangePassword(rpcmessages.UserChangePasswordArgs{Username: "admin", Password: testMiddleware.InitialAdminPassword(), NewPassword: "abc123def"})
+	t.Logf("%v", response)
```
Can be removed?

TheCharlatan replied:
done.

0xB10C commented on middleware/src/middleware_test.go:
```diff
@@ -236,106 +237,103 @@ func TestSetRootPassword(t *testing.T) {
-	testMiddleware := setupTestMiddleware()
+	testMiddleware := setupTestMiddleware(t)
 
 	/* test valid password change */
```
comment needs to be adjusted to code or removed

TheCharlatan replied:
adjusted.

Tomasvrba commented on middleware/src/rpcserver/rpcserver.go:
```diff
@@ -103,74 +110,146 @@ func (server *RPCServer) Serve() {
+// GetSetupStatus send the middleware's setup status as a SetupStatusResponse over rpc.
+func (server *RPCServer) GetSetupStatus(dummyArg bool, reply *rpcmessages.SetupStatusResponse) error {
+	*reply = server.middleware.SetupStatus()
+	log.Printf("sent erply %v: ", reply)
```
typo "reply"

TheCharlatan replied:
Fixed.

0xB10C commented on middleware/cmd/middleware/main.go:
```diff
@@ -53,7 +53,10 @@ func main() {
-	middleware := middleware.NewMiddleware(argumentMap, false)
+	middleware, err := middleware.NewMiddleware(argumentMap, false)
+	if err != nil {
+		log.Fatal("error starting the middleware, is redis connected? \nIf you are running the middleware outside of the base consider setting the redis mock flag to true.")
```
Please print at least the actual `err` error message! I think the error message is way to detailed. The suggestion will be outdated and confusing as soon as the another component initialized in the middlware returns an error.   

@Stadicus send me the picture of this log output below. He asked why the middleware can't connect to Redis, even if it's running. I'm guessing it's due to not having the Redis keys for the authentication setup in Redis set. It would be way easier to debug if it would say that the Redis key can't be found...
 
![image](https://user-images.githubusercontent.com/19157360/67686009-505a0c00-f996-11e9-8e9a-d84d5bfb3eba.png)

0xB10C replied:
I think it's an old image and has a newer middleware binary on it. I don't have access to the device. @Stadicus probably knows more.

TheCharlatan replied:
I'm also logging the error now.

TheCharlatan commented:
Rebased and squashed.
